File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-3.java
class Solution {
    public int characterReplacement(String s, int k) {
        int left = 0;
        int res = 0;
        int maxOcc = 0;
        int[] count = new int[26];

        for(int right=0; right<s.length(); right++) {
            count[s.charAt(right) - 'A']++;
            maxOcc = Math.max(maxOcc, count[s.charAt(right) - 'A']);

            if(right - left + 1 - maxOcc > k) {
                count[s.charAt(left) - 'A']--;
                left++;
            }

            res = Math.max(res, right - left + 1);
        }
        return res;
    }
}

File: Data Structures & Algorithms/eating-bananas/submission-3.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        left, right = 1, max(piles)

        while left < right:
            middle = (left + right) // 2
            hours = 0
            for pile in piles:
                hours += pile // middle
                if pile % middle != 0:
                    hours += 1
            if hours <= h:
                right = middle
            else:
                left = middle + 1
        return left